src/ConsumerBuilder.php

```
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Kafka\Consumer;
+
+use Closure;
+use Kafka\Consumer\Contracts\MessageDecoder;
+use Kafka\Consumer\Decoder\NoDecoding;
+use Kafka\Consumer\Entities\Config;
+use Kafka\Consumer\Entities\Config\Sasl;
+use Kafka\Consumer\MessageHandler\CallableConsumer;
+
+class ConsumerBuilder
+{
+    private $topics;
+    private $commit;
+    private $groupId;
+    private $handler;
+    private $maxMessages;
+    private $maxCommitRetries;
+    private $broker;
+    private $decoder;
+    private $saslConfig = null;
+    private $dlq;
+    private $securityProtocol;
+
+    private function __construct(array $topics)
+    {
+        $this->topics = $topics;
+        $this->commit = 1;
+        $this->groupId = 'php-kafka-' . uniqid();
+        $this->handler = function () {
+        };
+        $this->maxMessages = -1;
+        $this->maxCommitRetries = 6;
+        $this->broker = 'localhost:9092';
+        $this->decoder = new NoDecoding();
+        $this->securityProtocol = 'SASL_PLAINTEXT';
```

Some of these configs like brokers, security protocol and group id will cause the using application to not behave as intended if not set correctly. If we offer a more strict contract instead of dummy default values we can help the user catch config errors earlier.

---

What would you suggest for a more strict contract? Maybe do not assign dummy values and let it break while building the consumer? Or maybe adding some validation of required fields?

---

I think adding the required values to the create() signature is a good way to ensure the builder will have all it needs to work.

---

The only problem I see is loosing readability by adding too many parameters to the `create` method. But yeah, the brokers needs to be required and a dynamic group id in production might be problematic when scaling. However, for the securityProtocol, I think we can consider "PLAINTEXT" as the default one, at least this is the default for kafka as in the [docs](https://kafka.apache.org/0100/documentation.html#newconsumerconfigs).

So I suggest changing the create signature to: 

```php
public static function create(string $brokers, string $groupId, array $topics): self
{
    return new ConsumerBuilder($topics);
}
```